onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -359,6 +363,25 @@ Status CUDAExecutionProvider::SetComputeStream(void* stream) {
   return Status::OK();
 }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+void CUDAExecutionProvider::CaptureBegin()  {
+  ORT_THROW_IF_ERROR(Sync());
```

Comment:
Looking through the CUDA Graphs documentation, it seems like CPU operations are not captured and even they are they are elided during the replay. When we partition a model, there is chance that some ops be missing CUDA kernel support and these ops drop down to using the CPU EP. Should cuda graph only be captured when all the nodes of a graph have been assigned to the CUDA EP ?

Reply:
Good point!  Yes, cuda graph should be captured only when all the nodes are assigned to CPUDA EP. Do you think [OrtMemoryInfo](https://github.com/microsoft/onnxruntime/blob/baa17679221807ea332a7134c0d10b81adf4009e/onnxruntime/core/framework/sequential_execution_plan.h#L129) can be used to check these cases? For example if the device type is not GPU, then raise an error before graph capturing.

Reply:
This is a little tricky. We will have to do something like this - https://github.com/microsoft/onnxruntime/blob/ef7b4dc05cae1084546ff7caba048a2f908ac1d8/onnxruntime/core/session/inference_session.cc#L929. Check the node placements - and if anyone is placed on a non-CUDA EP and if the user has configured the CUDA EP to capture the graph, then fail the session load with a message that this model can't be used in the context of graph capture. Will think a bit more to see if there is a better solution...

Reply:
Added sample logic in InferenceSession::Initialize() in https://github.com/microsoft/onnxruntime/pull/10459 that disallows a session to initialize successfully if the user requests usage of cuda graphs and that can't be honored